hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -1742,6 +1742,71 @@ case class TableFilterIntervals(child: TableIR, intervals: IndexedSeq[Interval],
   }
 }
 
+case class TableGroupWithinPartitions(child: TableIR, n: Int) extends TableIR {
+  lazy val children: IndexedSeq[BaseIR] = Array(child)
+
+  lazy val rowCountUpperBound: Option[Long] = child.rowCountUpperBound
+
+  override lazy val typ: TableType = TableType(
+    child.typ.keyType ++ TStruct(("grouped_fields", TArray(child.typ.rowType))),
+    child.typ.key,
+    child.typ.globalType)
+
+  def copy(newChildren: IndexedSeq[BaseIR]): TableIR = {
+    val IndexedSeq(newChild: TableIR) = newChildren
+    TableGroupWithinPartitions(newChild, n)
+  }
+
+  override def execute(ctx: ExecuteContext): TableValue = {
+    val prev = child.execute(ctx)
+    val prevRVD = prev.rvd
+    val prevRowType = prev.rvd.typ.rowType
+    val prevKeyType = prev.rvd.typ.kType
+
+    val groupedElementsPType = PArray(prevRVD.typ.rowType, true)
+    val rowType = this.typ.rowType.physicalType
+    val newRVDType = prevRVD.typ.copy(rowType = rowType)
+    val keyIndices = child.typ.keyFieldIdx
+
+    val newRVD = prevRVD.mapPartitionsWithIndex(newRVDType, { (int, ctx, it) =>
+      val targetRegion = ctx.region
+
+      new Iterator[RegionValue] {
+        override def hasNext: Boolean = {
+          it.hasNext
+        }
+
+        override def next(): RegionValue = {
+          if (!hasNext)
+            throw new java.util.NoSuchElementException()
+
+          var regionValueArray = Array[RegionValue]()
+          var i = 0
+          while (it.hasNext && i != n) {
+            val nextRV = it.next()
+            regionValueArray = regionValueArray :+ nextRV
```

Comment:
this copies the array for each element. Instead, allocate the array with n elements, and just copy in the first `i` (whatever `i` is incremented to )